pkg/controller/bastion/options.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/gcp"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/extensions"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Options contains provider-related information required for setting up
+// a bastion instance. This struct combines precomputed values like the
+// bastion instance name with the IDs of pre-existing cloud provider
+// resources, like the Firewall name, subnet name etc.
+type Options struct {
+	Shoot               *gardencorev1beta1.Shoot
+	BastionInstanceName string
+	FirewallName        string
+	PublicIPs           []string
+	DiskName            string
+	Zone                string
+	Subnetwork          string
+	ProjectID           string
+	Network             string
+}
+
+// DetermineOptions determines the required information that are required to reconcile a Bastion on GCP. This
+// function does not create any IaaS resources.
+func DetermineOptions(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) (*Options, error) {
+	//Each resource name up to a maximum of 64 characters in GCP
+	//https://cloud.google.com/compute/docs/labeling-resources
+	name := cluster.Shoot.Name
```

Comment:
1. you probably wanted to link to https://cloud.google.com/compute/docs/naming-resources instead of https://cloud.google.com/compute/docs/labeling-resources or not?
2. the resource name limit is 63 characters and not 64
3. You changed
```go
name := cluster.ObjectMeta.Name
```
to 
```go
name := cluster.Shoot.Name
```
A shoot name could start with a number, however gcp has the limitation that the resource name must match the regular expression `[a-z]([-a-z0-9]*[a-z0-9])?` (see link above), which means that they must start with a lowercase letter. With your change the bastion creation would fail for shoots that start with a number.

4. Obviously you ran into an issue where you exceeded the 63 character limit for a resource. Which resource was it? Maybe we can optimize / shorten the resource name but still keep `name := cluster.ObjectMeta.Name`.

Reply:
Thanks let me know.  yes it should be 63 characters and naming-resources is better then labeling-resources 
```
	name := cluster.ObjectMeta.Name
	bastionInstanceName := fmt.Sprintf("%s-%s-bastion", name, bastion.Name)
	firewallName := fmt.Sprintf("%s-allow-ssh-access", bastionInstanceName)
```
In my local experiment, my project name is `ixxxxxx` the shoot name `ixxxxxx-gcp` then the `cluster.ObjectMeta.Name` will be like `shoot--ixxxxxx--ixxxxxx-gcp`
the bastion name in bastion manitest is `bastion-ixxxxxx`
then firewall rule name is like `shoot--ixxxxxx--ixxxxxx-gcp-bastion-ixxxxxx-bastion-allow-ssh-access` 68 characters.

maybe either name in bastion manifest use `ixxxxxx` in the future(60 characters)? or firewall rule name end with -ssh`-allow-ssh`  (61 characters). let me know WDYT

Reply:
committed in fbad5cfe22865b73a53ff0684b28bd8563af8e5d
I rename firewall rules `allow-ssh-access` to `allow-ssh`

Reply:
there is no length restriction for the bastion name (however `gardenctl-v2` creates a name like `cli-<8-random-character>`) in contrast to the shoot and project name. I guess sooner or later we will run again into this problem when someone creates a very long bastion name. We need to think about a different solution